cmd/appsctl/provision_app.go

```diff
@@ -0,0 +1,216 @@
+	var m *apps.Manifest
+	assets := []assetData{}
+
+	// Read all the files from zip archive
+	for _, file := range zipReader.File {
+		switch {
+		case strings.HasSuffix(file.Name, "manifest.json"):
+			var f io.ReadCloser
+			f, err = file.Open()
+			if err != nil {
+				return errors.Wrap(err, "can't open manifest.json file")
+			}
+			defer f.Close()
+
+			var data []byte
+			data, err = ioutil.ReadAll(f)
+			if err != nil {
+				return errors.Wrap(err, "can't read manifest.json file")
+			}
+			err = json.Unmarshal(data, &m)
+			if err != nil {
+				return errors.Wrapf(err, "can't unmarshal manifest.json file %s", string(data))
+			}
+
+		case strings.HasSuffix(file.Name, ".zip"):
+			var f io.ReadCloser
+			f, err = file.Open()
+			if err != nil {
+				return errors.Wrapf(err, "can't open file %s", file.Name)
+			}
+			defer f.Close()
+
+			bundleFunctions = append(bundleFunctions, functionInstallData{
+				name:   strings.TrimSuffix(file.Name, ".zip"),
+				bundle: f,
+			})
+			log.Debug("Found function bundle", "file", file.Name)
+
+		case strings.HasPrefix(file.Name, awsclient.StaticAssetsFolder):
+			assetName := strings.TrimPrefix(file.Name, awsclient.StaticAssetsFolder)
+			var f io.ReadCloser
+			f, err = file.Open()
+			if err != nil {
+				return errors.Wrapf(err, "can't open file %s", file.Name)
+			}
+			defer f.Close()
+
+			assets = append(assets, assetData{
+				name: assetName,
+				file: f,
+			})
+			log.Debug("Found static asset", "file", assetName)
+
+		default:
+			log.Info("Unknown file found in app bundle", "file", file.Name)
+		}
+	}
+
+	if m == nil {
+		return errors.New("no manifest found")
+	}
+
+	resFunctions := []functionInstallData{}
+
+	// Matching bundle functions to the functions listed in manifest
+	// O(n^2) code for simplicity
+	for _, bundleFunction := range bundleFunctions {
+		for _, manifestFunction := range m.Functions {
+			if strings.HasSuffix(bundleFunction.name, manifestFunction.Name) {
+				resFunctions = append(resFunctions, functionInstallData{
+					bundle:  bundleFunction.bundle,
+					name:    manifestFunction.Name,
+					handler: manifestFunction.Handler,
+					runtime: manifestFunction.Runtime,
+				})
+				continue
+			}
+		}
+	}
+
+	err = provisionAssets(awscli, m, assets)
+	if err != nil {
+		return errors.Wrapf(err, "can't provision assets of the app - %s", m.AppID)
+	}
+	err = provisionFunctions(awscli, m, resFunctions, shouldUpdate)
+	if err != nil {
+		return errors.Wrapf(err, "can't provision functions of the app - %s", m.AppID)
+	}
+	data, err := json.Marshal(m)
+	if err != nil {
+		return errors.Wrapf(err, "can't marshal manifest for app - %s", m.AppID)
+	}
+	buffer := bytes.NewBuffer(data)
+
+	bucket := awsclient.GenerateS3BucketNameWithDefaults("")
+	key := awsclient.GenerateManifestS3Name(m.AppID, m.Version)
+	if err := awscli.UploadS3(bucket, key, buffer); err != nil {
+		return errors.Wrapf(err, "can't upload manifest file for the app - %s", m.AppID)
+	}
```

Comment:
would be nice to have a method like `provisionManifest` or at least a comment. It would read nicely: provisionAssets, provisionFunctions, provisionManifest.